File: analysis/schedule_2026/scripts/schedule_leverage.py
"""Schedule Leverage Index — composite of non-opponent-quality schedule burdens.

Combines five environmental / structural dimensions into a single 0-100 score:
  1. Rest differential (negative = bad)
  2. Opponents-off-bye count (more = bad)
  3. Body-clock disadvantage games (more = bad)
  4. Travel miles (more = bad)
  5. Wind exposure mph-games (more = bad)

Each dimension is percentile-ranked within the league (100 = worst burden in that
dimension). Composite = equal-weighted average. Output: data.parquet with the
component scores + composite + leverage_rank.

This is intentionally SEPARATE from implied-wins SoS (opponent quality) — it's
the "non-opponent" lens: who has the toughest schedule environment to navigate,
holding opponent strength constant.
"""
from __future__ import annotations
from pathlib import Path
import sys

import pandas as pd

sys.path.insert(0, '/Users/nick/projects/dynasty-dashboard/analysis/schedule_2026')
from _shared import output, teams  # noqa: E402

ROOT = Path("/Users/nick/projects/dynasty-dashboard/analysis/schedule_2026")
SLUG = "schedule_leverage"


def pct_rank(series: pd.Series, *, ascending: bool) -> pd.Series:
    """Return 0-100 percentile. ascending=True means higher input value -> higher percentile."""
    r = series.rank(method="average", ascending=ascending, pct=True)
    return (r * 100).round(1)


def main() -> None:
    bye = pd.read_parquet(ROOT / "output/bye_leverage/data.parquet")
    bod = pd.read_parquet(ROOT / "output/body_clock/data.parquet")
    trv = pd.read_parquet(ROOT / "output/travel_miles/data.parquet")
    wnd = pd.read_parquet(ROOT / "output/wind_exposure/data.parquet")

    # rest_differential_sum: more NEGATIVE = worse, so ascending=False would put
    # most-negative at the top of percentile. We want "burden score" where high = bad,
    # so we use ascending=False (lower rest_diff -> higher rank).
    rest_burden  = pct_rank(bye["rest_differential_sum"], ascending=False)
    opps_off_bye = pct_rank(bye["opps_off_bye"], ascending=True)
    body_burden  = pct_rank(bod["disadvantage_games"], ascending=True)
    travel_burden= pct_rank(trv["total_miles"], ascending=True)
    wind_burden  = pct_rank(wnd["total_wind_exposure_mph_games"], ascending=True)

    base = pd.DataFrame({"team": bye["team"]})
    df = (base
          .merge(bye[["team", "rest_differential_sum", "opps_off_bye", "bye_week"]], on="team")
          .merge(bod[["team", "disadvantage_games"]], on="team")
          .merge(trv[["team", "total_miles", "intl_miles"]], on="team")
          .merge(wnd[["team", "total_wind_exposure_mph_games"]], on="team"))

    df["rest_burden_pct"]   = df["team"].map(dict(zip(bye["team"], rest_burden)))
    df["off_bye_burden_pct"]= df["team"].map(dict(zip(bye["team"], opps_off_bye)))
    df["body_clock_burden_pct"] = df["team"].map(dict(zip(bod["team"], body_burden)))
    df["travel_burden_pct"] = df["team"].map(dict(zip(trv["team"], travel_burden)))
    df["wind_burden_pct"]   = df["team"].map(dict(zip(wnd["team"], wind_burden)))

    component_cols = ["rest_burden_pct", "off_bye_burden_pct",
                      "body_clock_burden_pct", "travel_burden_pct", "wind_burden_pct"]
    df["leverage_index"] = df[component_cols].mean(axis=1).round(1)

    # Identify each team's top burden — i.e., which dimension is doing the most damage
    pretty = {
        "rest_burden_pct":       "rest",
        "off_bye_burden_pct":    "off-bye opps",
        "body_clock_burden_pct": "body-clock",
        "travel_burden_pct":     "travel",
        "wind_burden_pct":       "wind",
    }
    df["top_burden"] = df[component_cols].idxmax(axis=1).map(pretty)
    df["top_burden_pct"] = df[component_cols].max(axis=1).round(1)

    df["division"] = df["team"].map(teams.DIVISIONS)
    df["leverage_rank"] = df["leverage_index"].rank(ascending=False, method="min").astype(int)
    df = df.sort_values("leverage_index", ascending=False).reset_index(drop=True)

    output.write_data(SLUG, df)

    print(df[["leverage_rank","team","leverage_index","top_burden","top_burden_pct"]].to_string(index=False))


if __name__ == "__main__":
    main()

File: analysis/schedule_2026/scripts/viz_r/schedule_leverage.R
## Schedule Leverage Index — composite of 5 non-opponent-quality burdens.
## Horizontal bar of leverage_index (0-100) with each team's top burden labeled.

source("/Users/nick/projects/dynasty-dashboard/analysis/schedule_2026/scripts/viz_r/_theme.R")

d <- read_parquet(file.path(PROJ_ROOT, "output/schedule_leverage/data.parquet"))

d <- d |>
  dplyr::arrange(leverage_index) |>
  dplyr::mutate(team = factor(team, levels = team))

league_avg <- mean(d$leverage_index)
n_rows <- length(levels(d$team))

p <- ggplot(d, aes(x = leverage_index, y = team)) +
  geom_col(width = 0.78, colour = NA, fill = PAL$slate) +
  geom_text(aes(label = sprintf("%.0f  ·  driver: %s (%.0f)",
                                leverage_index, top_burden, top_burden_pct),
                hjust = -0.1),
            size = 3.4, colour = PAL$text) +
  scale_x_continuous(limits = c(0, 110),
                     breaks = seq(0, 100, by = 20),
                     expand = expansion(0)) +
  geom_vline(xintercept = league_avg, linetype = "31",
             colour = PAL$muted, linewidth = 0.5) +
  annotate("text", x = league_avg, y = n_rows + 0.8,
           label = sprintf("league avg (%.0f)", league_avg),
           hjust = 0.5, colour = PAL$muted, size = 3.4) +
  coord_cartesian(clip = "off") +
  labs(
    title    = "2026 Schedule Leverage Index",
    subtitle = paste0(
      "Composite of 5 non-opponent-quality schedule burdens, each percentile-ranked within the league ",
      "and equal-weighted: <b>rest differential · opponents-off-bye · body-clock disadvantage · travel miles · wind exposure</b>. ",
      "Higher score = tougher schedule environment to navigate, independent of opponent strength."
    ),
    x        = "Leverage index (0-100, higher = more environmental burden)",
    y        = NULL,
    caption  = "Sources: bye_leverage · body_clock · travel_miles · wind_exposure (all 2026 schedule-derived) · logos: nflplotR"
  ) +
  theme_sched(14) +
  theme(
    axis.text.y         = nflplotR::element_nfl_logo(size = 1.0),
    axis.ticks.y        = element_blank(),
    panel.grid.major.y  = element_blank()
  )

save_chart(p, file.path(PROJ_ROOT, "output/schedule_leverage/chart.png"),
           width = 16, height = 13)
